text/3804-cfg-alias.md

```diff
@@ -0,0 +1,165 @@
+duplicated, and giving it a name provides an easy way to show what a group of
+configuration is intended to represent.
+
+Something to this effect can be done using build scripts. This requires reading
+various Cargo environment variables and potentially doing string manipulation
+(for splitting target features), so it is often inconvenient enough to not be
+worth doing. Allowing aliases to be defined within the crate and with the same
+syntax as the `cfg` itself makes this much easier.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+There is a new crate-level attribute that takes a name and a `cfg` predicate:
+
+```rust
+#![cfg_alias(some_alias, predicate)]
+```
+
+`predicate` can be anything that usually works within `#[cfg(...)]`, including
+`all`, `any`, and `not`.
+
+Once an alias is defined, `name` can be used as if it had been passed via
+`--cfg`:
+
+```rust
+#[cfg(some_alias)]
+struct Foo { /* ... */ }
+
+#[cfg(not(some_alias))]
+struct Foo { /* ... */ }
+
+#[cfg(all(some_alias, target_os = "linux"))]
+fn bar() { /* ... */ }
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The new crate-level attribute is introduced:
+
+```text
+CfgAliasAttribute:
+    cfg_alias(IDENTIFIER, ConfigurationPredicate)
+```
+
+The identifier is added to the `cfg` namespace. It must not conflict with any
+builtin configuration names, or with those passed via `--cfg`.
+
+Once defined, the alias can be used as a regular predicate.
+
+The alias is only usable after it has been defined. For example, the following
+will emit an unknown configuration lint:
+
+```rust
+#![cfg_attr(some_alias, some_attribute)]
+// warning: unexpected_cfgs
+//
+// The lint could mention that `some_alias` was found in the
+// crate but is not available here.
+
+#![cfg_alias(some_alias, true)]
+```
+
+RFC Question:
+
+Two ways to implement this are with (1) near-literal substitution, or (2)
+checking whether the alias should be set or not at the time it is defined. Is
+there any user-visible behavior that would make us need to specify one or the
+other?
+
+If we go with the first option, we should limit to a single expansion to avoid
+recursing (as is done for `#define` in C).
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- This does not support more general attribute aliases, such as
+  `#![alias(foo = derive(Clone, Copy, Debug, Default)`. This seems better suited
+  for something like `declarative_attribute_macros` in [RFC3697].
+
+[RFC3697]: https://github.com/rust-lang/rfcs/pull/3697
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The syntax `cfg_alias(name, predicate)` was chosen for similarity with
+  `cfg_attr(predicate, attributes)`. Alternatives include:
+  - `cfg_alias(name = predicate)`
+- It may be possible to have `#[cfg_alias(...)]` work as an outer macro and only
+  apply to a specific scope. This likely is not worth the complexity.
```

Comment:
I think it would be more consistent to let it work at a module scope, if nothing else. I agree that the most important use cases are at the crate level, and if anything more does add a lot of complexity, it might not be worth it.

Reply:
I wouldn't mind having it as well but didn't know if it would require more invasive changes to config handling. I'll update to include this, but indicate that we may choose to drop or defer that portion of the feature based on implementation complexity.

Reply:
Added this at https://github.com/tgross35/rfcs/blob/cfg-alias/text/3804-cfg-alias.md#cfg_alias-in-non-crate-attributes